src/hotspot/share/compiler/methodMatcher.cpp

```diff
@@ -224,7 +224,11 @@ static MethodMatcher::Mode check_mode(char name[], const char*& error_msg) {
       return MethodMatcher::Any;
     }
     match |= MethodMatcher::Suffix;
-    memmove(name, name + 1, strlen(name + 1) + 1);
+PRAGMA_DIAG_PUSH
+PRAGMA_STRINGOP_OVERREAD_IGNORED
+    // This code can incorrectly cause a "stringop-overread" warning with gcc
+     memmove(name, name + 1, strlen(name + 1) + 1);
+PRAGMA_DIAG_POP
```

Comment:
This function has 4 `strlen(name)` expressions.  It could be changed to have
only one, and then update the value as it proceeds. I wonder if changing it
that way might dodge whatever is confusing ASAN, as well as (arguably) making
the function a bit clearer.  I've no idea how important (or not) reducing the
number of strlen calls might be for performance.

So something like this (not tested):
```
static MethodMatcher::Mode check_mode(char name[], const char*& error_msg) {
  int match = MethodMatcher::Exact;
  size_t len = strlen(name);
  if (name[0] == '*') {
    if (len == 1) {
      return MethodMatcher::Any;
    }
    match |= MethodMatcher::Suffix;
    memmove(name, name + 1, len); // Include terminating nul in move.
    len -= 1;
  }

  if (len > 0 && name[len - 1] == '*') {
    match |= MethodMatcher::Prefix;
    name[--len] = '\0';
  }

  if (len == 0) {
    error_msg = "** Not a valid pattern";
    return MethodMatcher::Any;
  }

  if (strstr(name, "*") != nullptr) {
    error_msg = " Embedded * not allowed";
    return MethodMatcher::Unknown;
  }
  return (MethodMatcher::Mode)match;
}
```

Reply:
Thanks your advice. The performance impact maybe ignore, since this function used for parser `-XX:CompileCommand`. I will try this change can avoid gcc warning or not with -fsanitize=undefined -fsanitize=address.

Reply:
The four strlen calculation has been combine to one. Thanks.